beluga/include/beluga/views/sample.hpp

```diff
@@ -0,0 +1,210 @@
+ *  [RandomNumberDistribution](https://en.cppreference.com/w/cpp/named_req/RandomNumberDistribution).
+ * \tparam URNG A random number generator that satisfies the
+ *  [UniformRandomBitGenerator](https://en.cppreference.com/w/cpp/named_req/UniformRandomBitGenerator)
+ *  requirements.
+ */
+template <class Range, class Distribution, class URNG = typename ranges::detail::default_random_engine>
+struct sample_view : public ranges::view_facade<sample_view<Range, Distribution, URNG>, ranges::infinite> {
+ public:
+  /// Default constructor.
+  sample_view() = default;
+
+  /// Construct the view from an existing range.
+  /**
+   * \param range The range to be adapted.
+   * \param distribution The random number distribution to use to sample elements.
+   * \param engine The random number generator object.
+   */
+  constexpr sample_view(Range range, Distribution distribution, URNG& engine = ranges::detail::get_random_engine())
+      : range_{std::move(range)},
+        distribution_{std::make_shared<Distribution>(std::move(distribution))},
+        engine_{std::addressof(engine)} {
+    assert(distribution_->min() == 0);
+    assert(distribution_->max() == ranges::size(range_) - 1);
+  }
+
+ private:
+  // `ranges::range_access` needs access to the cursor members.
+  friend ranges::range_access;
+
+  static_assert(ranges::sized_range<Range>);
+  static_assert(ranges::random_access_range<Range>);
+  static_assert(std::is_same_v<typename Distribution::result_type, ranges::range_difference_t<Range>>);
+
+  /// Cursor class that handles the iteration logic.
+  struct cursor {
+   public:
+    /// Default constructor.
+    cursor() = default;
+
+    /// Construct a cursor from the parent view elements.
+    constexpr cursor(std::shared_ptr<Distribution> distribution, URNG* engine, ranges::iterator_t<Range> first)
+        : distribution_{std::move(distribution)}, engine_{engine}, first_{std::move(first)} {
+      assert(distribution_ != nullptr);
+      next();
+    }
+
+    /// Access the current iterator.
+    [[nodiscard]] decltype(auto) read() const noexcept(noexcept(*this->it_)) { return *it_; }
+
+    /// Position the current iterator.
+    void next() { it_ = first_ + (*distribution_)(*engine_); }
+
+   private:
+    std::shared_ptr<Distribution> distribution_;
+    URNG* engine_;
+    ranges::iterator_t<Range> first_;
+    ranges::iterator_t<Range> it_;
+  };
+
+  /// Return the cursor for the begin iterator.
+  [[nodiscard]] constexpr auto begin_cursor() { return cursor{distribution_, engine_, ranges::begin(range_)}; }
+
+  /// Return an unreachable sentinel since this is an infinite range.
+  [[nodiscard]] constexpr auto end_cursor() const noexcept { return ranges::unreachable_sentinel_t{}; }
+
+  Range range_;
+
+  // Using a shared resource as the current version of range-v3 does not support move-only views.
+  // This is to ensure the O(1) (cheaply) copyable semantic requirement for views.
+  std::shared_ptr<Distribution> distribution_;
```

Comment:
@nahueespinosa thought: we could escape dynamic allocation here by relying on the fact all STL distributions support near copy construction via parameter set getters and the corresponding constructor (see `param` methods). Unclear what the complexity of distribution instantiation is.

Reply:
What I'm trying to avoid here is copying the vector of weights that `discrete_distribution` parameters hold:
https://github.com/gcc-mirror/gcc/blob/cddaff46adfd831d4836f29dc5e21f73eaed37b3/libstdc%2B%2B-v3/include/bits/random.h#L5499

Reply:
Yeah, I know that's the downside, and in this case, you'll heap allocate either way. My point is, near copy constructing distributions leaves the door open to stack allocated (and real-time friendly) implementations.

Anyways, I'm over-optimizing. Nevermind.